File: 2024/day_3/Part2.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Part2 {
  private static String readInput(String[] args) throws IOException {
    if (args.length == 0) return new String(System.in.readAllBytes());
    return new String(Files.readAllBytes(Paths.get(args[0])));
  }

  public static String substring(String s, int start, int end) {
    return s.substring(start, end > s.length() ? s.length() : end);
  }

  public static void main(String[] args) throws IOException {
    String input = readInput(args);
    int sum = 0;
    boolean shouldDo = true;
    for (int i = 0; i < input.length(); i++) {
      if (substring(input, i, i + "don't()".length()).equals("don't()")) {
        shouldDo = false;
        i += "don't()".length() - 1;
        continue;
      }
      if (substring(input, i, i + "do()".length()).equals("do()")) {
        shouldDo = true;
        i += "do()".length() - 1;
        continue;
      }
      if (substring(input, i, i + "mul(".length()).equals("mul(")) {
        int nextClose = input.indexOf(")", i + "mul(".length());
        String sub = substring(input, i + "mul(".length(), nextClose);
        String[] parts = sub.split(",");
        int a, b;
        try {
          a = Integer.parseInt(parts[0]);
          b = Integer.parseInt(parts[1]);
        } catch (Exception e) {
          i += "mul(".length() - 1;
          continue;
        }
        sum += shouldDo ? a * b : 0;
        i = nextClose;
      }
    }
    System.out.println(sum);

  }
}

File: 2024/day_3/part2.ts
// https://adventofcode.com/2024/day/1

import { readInput } from "../../utils/utils";

const input = await readInput();

let sum = 0;
let shouldDo = true;
for (let i = 0; i < input.length; i++) {
  if (input.substring(i, i + "don't()".length) === "don't()") {
    shouldDo = false;
    i += "don't()".length - 1;
    continue;
  }
  if (input.substring(i, i + "do()".length) === "do()") {
    shouldDo = true;
    i += "do()".length - 1;
    continue;
  }
  if (input.substring(i, i + "mul(".length) === "mul(") {
    const nextClose = input.indexOf(")", i + "mul(".length);
    const match = input.substring(i, nextClose + 1).match(/mul\((\d+),(\d+)\)/);
    if (!match) {
      i += "mul(".length - 1;
      continue;
    }
    sum += shouldDo ? +match[1] * +match[2] : 0;
    i = nextClose;
  }
}
console.log(sum);
